File: lib/models/item/comment.dart
import 'package:hacki/models/item/item.dart';

class Comment extends Item {
  Comment({
    required super.id,
    required super.time,
    required super.parent,
    required super.score,
    required super.by,
    required super.text,
    required super.kids,
    required super.dead,
    required super.deleted,
    required super.hidden,
    required this.level,
    required this.isFromCache,
    this.isHiddenByUser = false,
    this.isCollapsedByUser = false,
    this.isNew = false,
  }) : super(
          descendants: 0,
          parts: <int>[],
          title: '',
          url: '',
          type: '',
        );

  Comment.fromJson(super.json, {this.level = 0})
      : isFromCache = json['fromCache'] == true,
        isHiddenByUser = false,
        isCollapsedByUser = false,
        isNew = false,
        super.fromJson();

  Comment.fromJsonWithCollapsedStateOnly(super.json, {this.level = 0})
      : isFromCache = json['fromCache'] == true,
        isHiddenByUser = json['isHiddenByUser'] == true,
        isCollapsedByUser = json['isCollapsedByUser'] == true,
        isNew = false,
        super.fromJson();

  final int level;
  final bool isHiddenByUser;
  final bool isCollapsedByUser;
  final bool isFromCache;
  final bool isNew;

  String get metadata => '''by $by $timeAgo''';

  bool get isRoot => level == 0;

  Comment copyWith({
    int? level,
    int? kid,
    bool? hidden,
    bool? isHiddenByUser,
    bool? isCollapsedByUser,
    bool? isNew,
  }) {
    return Comment(
      id: id,
      time: time,
      parent: parent,
      score: score,
      by: by,
      text: text,
      kids: kid == null ? kids : <int>[...kids, kid],
      dead: dead,
      deleted: deleted,
      hidden: hidden ?? this.hidden,
      level: level ?? this.level,
      isFromCache: isFromCache,
      isHiddenByUser: isHiddenByUser ?? this.isHiddenByUser,
      isCollapsedByUser: isCollapsedByUser ?? this.isCollapsedByUser,
      isNew: isNew ?? this.isNew,
    );
  }

  Comment copyWithOnlyCollapseState() {
    return Comment(
      id: id,
      time: time,
      parent: parent,
      score: score,
      by: by,
      text: '',
      kids: kids,
      dead: dead,
      deleted: deleted,
      hidden: hidden,
      level: level,
      isFromCache: isFromCache,
      isHiddenByUser: isHiddenByUser,
      isCollapsedByUser: isCollapsedByUser,
    );
  }

  Map<String, dynamic> toJsonWithOnlyCollapseState() {
    return <String, dynamic>{
      ...super.toJson(full: false),
      'isHiddenByUser': isHiddenByUser,
      'isCollapsedByUser': isCollapsedByUser,
    };
  }

  @override
  bool? get stringify => false;

  @override
  List<Object?> get props => <Object?>[
        ...super.props,
        isHiddenByUser,
        isCollapsedByUser,
        isNew,
      ];
}

File: ios/StoryWidget/Models/Comment.swift
public struct Comment: Item {
    public let id: Int
    public let parent: Int?
    public let text: String?
    public let type: String?
    public let by: String?
    public let time: Int
    public let kids: [Int]?
    public let level: Int?
    public var metadata: String {
        if let count = kids?.count, count != 0 {
            return "\(count) cmt\(count > 1 ? "s":"") | \(timeAgo) by \(by.orEmpty)"
        } else {
            return "\(timeAgo) by \(by.orEmpty)"
        }
    }
    
    /// Values below will always be nil for `Comment`.
    public let title: String?
    public let url: String?
    public let descendants: Int?
    public let score: Int?


    init(id: Int, parent: Int?, text: String?, by: String?, time: Int, kids: [Int]? = [Int](), level: Int? = 0) {
        self.id = id
        self.parent = parent
        self.text = text
        self.by = by
        self.time = time
        self.kids = kids
        self.level = level
        self.type = "comment"
        self.title = nil
        self.url = nil
        self.descendants = nil
        self.score = nil
    }

    // Empty initializer
    init() {
        self.init(id: 0, parent: 0, text: "", by: "", time: 0)
    }

    public func copyWith(text: String? = nil, level: Int? = nil) -> Comment {
        Comment(id: id, 
                parent: parent,
                text: text ?? self.text,
                by: by,
                time: time,
                kids: kids ?? [Int](),
                level: level ?? self.level)
    }
}
